File: packages/client/src/direct.ts
/**
 * @fileoverview Direct Transport for zero-overhead in-process command execution
 *
 * This transport enables co-located agents to execute commands directly
 * without any transport overhead (no JSON-RPC, no IPC, no network).
 *
 * @example
 * ```typescript
 * import { createDirectClient } from '@lushly-dev/afd-client';
 * import { createDirectRegistry } from '@lushly-dev/afd-server';
 * import { commands } from '@my-app/commands';
 *
 * // Zod validation, middleware and expose.agent checks, in process
 * const client = createDirectClient(createDirectRegistry(commands));
 * const result = await client.call('todo-create', { title: 'Fast!' }, { timeout: 5000 });
 * ```
 */

import type {
	CommandContext,
	CommandMiddleware,
	CommandResult,
	HandoffResult,
	PipelineRequest,
	PipelineResult,
	PipelineStep,
} from '@lushly-dev/afd-core';
import {
	executePipeline,
	executionFailure,
	failure,
	truncateName,
	validationError,
} from '@lushly-dev/afd-core';
import { runWithTimeout } from './direct-timeout.js';
import { validateInput } from './direct-validation.js';
import type {
	HandoffConnection,
	HandoffConnectionOptions,
	ReconnectingHandoffConnection,
	ReconnectionOptions,
} from './handoff.js';
import { connectHandoff, createReconnectingHandoff } from './handoff.js';
import type { UnknownToolError } from './unknown-tool.js';
import { unknownToolFailure } from './unknown-tool.js';

// Re-export extracted modules for backward compatibility
export { DirectTransport } from './direct-transport.js';
export type { CommandDefinition, CommandParameter } from './direct-validation.js';
export { isUnknownToolError, type UnknownToolError } from './unknown-tool.js';

// Option, context and registry types live in direct-types.ts
import type { DirectCallContext, DirectClientOptions, DirectRegistry } from './direct-types.js';

export type { DirectCallContext, DirectClientOptions, DirectRegistry } from './direct-types.js';

/**
 * Generate a unique trace ID.
 */
function generateTraceId(): string {
	return `trace-${Date.now()}-${Math.random().toString(36).slice(2, 9)}`;
}

/**
 * DirectClient for zero-overhead in-process command execution.
 *
 * Use this when you don't need the full McpClient features like
 * reconnection, events, etc. This provides the minimal API for
 * command execution with optional validation and context propagation.
 *
 * @example
 * ```typescript
 * import { isSuccess } from '@lushly-dev/afd-core';
 * import { createDirectClient, isUnknownToolError } from '@lushly-dev/afd-client';
 * import { createDirectRegistry } from '@lushly-dev/afd-server';
 * import { commands } from '@my-app/commands';
 *
 * const client = createDirectClient(createDirectRegistry(commands));
 *
 * // Type-safe command execution
 * const result = await client.call<Todo>('todo-create', { title: 'Test' });
 *
 * if (isUnknownToolError(result)) {
 *   console.log(result.data?.hint); // e.g. "Did you mean 'todo-create'?"
 * } else if (isSuccess(result)) {
 *   console.log(result.data.id); // result is CommandResult<Todo>
 * }
 * ```
 *
 * @example With context propagation
 * ```typescript
 * const client = createDirectClient(registry, {
 *   source: 'my-agent',
 *   debug: true,
 * });
 *
 * // Context is propagated to command handlers
 * const result = await client.call('command', args, { traceId: 'custom-trace' });
 * ```
 */
export class DirectClient {
	private readonly options: Required<
		Omit<DirectClientOptions, 'source' | 'middleware' | 'allow'>
	> & {
		source?: string;
		middleware: CommandMiddleware[];
		allow?: (commandName: string) => boolean;
	};

	constructor(
		private readonly registry: DirectRegistry,
		options: DirectClientOptions = {}
	) {
		this.options = {
			source: options.source,
			debug: options.debug ?? false,
			validateInputs: options.validateInputs ?? true,
			middleware: options.middleware ?? [],
			allow: options.allow,
		};
	}

	/** Whether the `allow` option lets this client call a command. A throwing predicate denies. */
	private isAllowed(name: string): boolean {
		if (!this.options.allow) return true;
		try {
			return this.options.allow(name) === true;
		} catch {
			return false;
		}
	}

	/**
	 * Call a command and return a CommandResult.
	 *
	 * This is the most efficient path - direct registry access
	 * with zero transport overhead.
	 *
	 * @param name - Command name
	 * @param args - Command arguments
	 * @param context - Optional context for tracing/cancellation
	 * @returns Command result
	 */
	async call<T = unknown>(
		name: string,
		args?: Record<string, unknown>,
		context?: DirectCallContext
	): Promise<CommandResult<T> | CommandResult<UnknownToolError>> {
		const startTime = this.options.debug ? performance.now() : 0;
		const traceId = context?.traceId ?? generateTraceId();

		this.debug(`[${traceId}] Calling ${name}`, args);

		if (!this.isAllowed(name)) {
			this.debug(`[${traceId}] Not allowed: ${name}`);
			return failure<T>({
				code: 'COMMAND_NOT_ALLOWED',
				message: `Command '${truncateName(name)}' is not allowed for this client`,
				suggestion: 'Call one of the commands returned by listCommandNames()',
				retryable: false,
			});
		}

		// Check if the command exists - return structured error if not
		if (!this.registry.hasCommand(name)) {
			this.debug(`[${traceId}] Unknown command: ${name}`);
			// The UnknownToolError is the data, so agents get structured recovery information.
			return unknownToolFailure(name, this.listCommandNames());
		}

		// Input validation (if registry supports getCommand)
		if (this.options.validateInputs && this.registry.getCommand) {
			const command = this.registry.getCommand(name);
			if (command) {
				const issues = validateInput(args, command.parameters);
				if (issues) {
					this.debug(`[${traceId}] Validation failed:`, issues);

					return failure(
						validationError(
							`Invalid input for '${name}': ${issues.map((i) => i.message).join('; ')}`,
							{ issues }
						)
					) as CommandResult<T>;
				}
			}
		}

		// Build command context
		const commandContext: CommandContext = {
			traceId,
			...context,
		};

		// Add source if configured
		if (this.options.source) {
			commandContext.source = this.options.source;
		}

		// Execute the command (with middleware if configured), enforcing context.timeout
		const result = await runWithTimeout<T>(name, context?.timeout, context?.signal, (signal) => {
			if (signal) commandContext.signal = signal;
			return this.execute<T>(name, args, commandContext);
		});

		if (this.options.debug) {
			const duration = performance.now() - startTime;
			this.debug(`[${traceId}] Completed in ${duration.toFixed(3)}ms`, {
				success: result.success,
			});
		}

		return result;
	}

	/**
	 * Run the registry through the client middleware (none: the zero-overhead path). A registry,
	 * handler or middleware that throws gives a `COMMAND_EXECUTION_ERROR` failure (without the
	 * exception text), as `createDirectRegistry()` and the MCP server do, instead of a rejection.
	 */
	private async execute<T>(
		name: string,
		args: Record<string, unknown> | undefined,
		commandContext: CommandContext
	): Promise<CommandResult<T>> {
		let next: () => Promise<CommandResult> = () =>
			this.registry.execute<T>(name, args, commandContext);
		for (let i = this.options.middleware.length - 1; i >= 0; i--) {
			const mw = this.options.middleware[i];
			if (!mw) continue;
			const currentNext = next;
			next = () => mw(name, args ?? {}, commandContext, currentNext);
		}
		try {
			return (await next()) as CommandResult<T>;
		} catch (error) {
			this.debug(`Command '${truncateName(name)}' threw:`, error);
			return executionFailure(error);
		}
	}

	/**
	 * List available commands (those the `allow` option permits).
	 */
	listCommands(): Array<{ name: string; description: string }> {
		return this.registry.listCommands().filter((command) => this.isAllowed(command.name));
	}

	/**
	 * List command names (those the `allow` option permits).
	 */
	listCommandNames(): string[] {
		return this.registry.listCommandNames().filter((name) => this.isAllowed(name));
	}

	/**
	 * Check if a command exists and the `allow` option permits it.
	 */
	hasCommand(name: string): boolean {
		return this.isAllowed(name) && this.registry.hasCommand(name);
	}

	/**
	 * Get the source identifier for this client.
	 */
	getSource(): string | undefined {
		return this.options.source;
	}

	/**
	 * Connect to a handoff endpoint using the appropriate protocol handler.
	 *
	 * @param handoff - The handoff result from a command
	 * @param options - Connection options and callbacks
	 * @returns A promise that resolves to a HandoffConnection
	 * @throws Error if no handler is registered for the protocol
	 */
	async connectHandoff(
		handoff: HandoffResult,
		options: HandoffConnectionOptions = {}
	): Promise<HandoffConnection> {
		return connectHandoff(handoff, options);
	}

	/**
	 * Create a reconnecting handoff connection with automatic retry logic.
	 *
	 * @param handoff - The initial handoff result
	 * @param options - Reconnection options and callbacks
	 * @returns A promise that resolves to a ReconnectingHandoffConnection
	 */
	async createReconnectingHandoff(
		handoff: HandoffResult,
		options: ReconnectionOptions = {}
	): Promise<ReconnectingHandoffConnection> {
		return createReconnectingHandoff(this, handoff, options);
	}

	/**
	 * Execute a pipeline of commands, chaining outputs to inputs.
	 *
	 * Pipelines allow declarative composition of commands where the output
	 * of one step flows into the next. Supports variable resolution
	 * ($prev, $first, $steps[n], $steps.alias, $input), conditional execution,
	 * and aggregated metadata (confidence, reasoning, warnings). Resolution follows
	 * `spec/pipeline-variables.md`: other `$` strings are literals, `$$` escapes a
	 * literal `$`, and unresolved references are omitted (or `null` in arrays).
	 *
	 * **Behavior change:** `$input` resolves to the request's own `input` field.
	 * It used to resolve to `context`, so a step could copy trace IDs, auth or any
	 * custom call-context value into a command input. `context` still reaches every
	 * command's context, but no reference can read it. Pass pipeline data as
	 * `pipe({ input, steps })` instead.
	 *
	 * Each step's data is copied as it is recorded and as it is resolved, so a
	 * handler that mutates its input cannot change another step's data.
	 *
	 * @param request - Pipeline request (with optional `input`) or array of steps
	 * @param context - Optional call context passed to each command; not visible to `$input`
	 * @returns Pipeline result with final data and aggregated metadata
	 *
	 * @example Basic pipeline
	 * ```typescript
	 * const result = await client.pipe([
	 *   { command: 'user-get', input: { id: 123 }, as: 'user' },
	 *   { command: 'order-list', input: { userId: '$prev.id' } },
	 *   { command: 'order-total', input: { orders: '$prev' } }
	 * ]);
	 *
	 * console.log(result.data); // Total from last step
	 * console.log(result.metadata.confidence); // Minimum confidence across steps
	 * ```
	 *
	 * @example Conditional execution
	 * ```typescript
	 * const result = await client.pipe([
	 *   { command: 'user-get', input: { id: 123 }, as: 'user' },
	 *   {
	 *     command: 'premium-features',
	 *     input: { userId: '$steps.user.id' },
	 *     when: { $eq: ['$steps.user.tier', 'premium'] }
	 *   }
	 * ]);
	 * ```
	 *
	 * @example Pipeline input
	 * ```typescript
	 * const result = await client.pipe({
	 *   input: { userId: 123 },
	 *   steps: [{ command: 'order-list', input: { userId: '$input.userId' } }],
	 * });
	 * ```
	 */
	async pipe<T = unknown>(
		request: PipelineRequest | PipelineStep[],
		context?: DirectCallContext
	): Promise<PipelineResult<T>> {
		const traceId = context?.traceId ?? generateTraceId();
		const pipelineRequest: PipelineRequest = Array.isArray(request) ? { steps: request } : request;
		this.debug(`[${traceId}] Starting pipeline`);
		let stepIndex = 0;
		return executePipeline(
			pipelineRequest,
			(command, input, commandContext) =>
				this.call(command, input as Record<string, unknown>, {
					...context,
					...commandContext,
					traceId: `${traceId}-step-${stepIndex++}`,
				}) as Promise<CommandResult>,
			{ ...context, traceId }
		) as Promise<PipelineResult<T>>;
	}

	/**
	 * Debug logging helper.
	 */
	private debug(message: string, data?: unknown): void {
		if (this.options.debug) {
			console.log(`[DirectClient] ${message}`, data ?? '');
		}
	}
}

/**
 * Create a new DirectClient for zero-overhead command execution.
 *
 * This is the recommended way to create a DirectClient. It provides
 * the same API as McpClient.call() but with ~100x faster execution
 * for co-located modules.
 *
 * @param registry - The command registry to execute against
 * @param options - Optional configuration for validation, context, and debugging
 * @returns A new DirectClient instance
 *
 * @example
 * ```typescript
 * import { createDirectClient } from '@lushly-dev/afd-client';
 * import { createDirectRegistry } from '@lushly-dev/afd-server';
 * import { commands } from './commands';
 *
 * const client = createDirectClient(createDirectRegistry(commands), {
 *   source: 'garden-api',
 *   allow: (name) => name.startsWith('plant-'),
 * });
 * const result = await client.call('plant-get', { id: 'tomato-123' });
 * // ~0.03-0.1ms latency vs ~2-10ms for MCP
 * ```
 */
export function createDirectClient(
	registry: DirectRegistry,
	options?: DirectClientOptions
): DirectClient {
	return new DirectClient(registry, options);
}

File: python/src/afd/direct.py
"""Direct Transport for zero-overhead in-process command execution.

This module enables co-located agents to execute commands directly
without any transport overhead (no JSON-RPC, no IPC, no network).

Example:
    >>> from afd.direct import DirectClient, create_registry
    >>> from my_app import registry
    >>>
    >>> client = DirectClient(registry)
    >>> result = await client.call('todo.create', {'title': 'Fast!'})
    >>> # ~0.1ms latency vs 10-100ms for MCP

Pipeline Example:
    >>> result = await client.pipe([
    ...     {'command': 'user.get', 'input': {'id': 123}, 'as': 'user'},
    ...     {'command': 'order.list', 'input': {'user_id': '$user.id'}},
    ... ])
"""

from __future__ import annotations

import time
import uuid
from abc import abstractmethod
from dataclasses import dataclass, field
from typing import (
    Any,
    Callable,
    Dict,
    List,
    Optional,
    Protocol,
    Sequence,
    TypeVar,
    Union,
    runtime_checkable,
)

from afd.core.result import CommandResult, failure, success
from afd.core.errors import not_found_error, validation_error


# ═══════════════════════════════════════════════════════════════════════════════
# TYPES
# ═══════════════════════════════════════════════════════════════════════════════

T = TypeVar("T")


@dataclass
class CommandContext:
    """Context for command execution."""
    
    trace_id: str
    source: Optional[str] = None
    timeout: Optional[float] = None
    extra: Dict[str, Any] = field(default_factory=dict)


@dataclass
class CommandInfo:
    """Information about a registered command."""
    
    name: str
    description: str


@dataclass
class CommandParameter:
    """Command parameter definition for validation."""
    
    name: str
    type: str  # 'string', 'number', 'boolean', 'object', 'array'
    description: str = ""
    required: bool = True
    default: Any = None
    enum: Optional[List[Any]] = None


@dataclass
class CommandDefinition:
    """Full command definition with parameters."""
    
    name: str
    description: str
    parameters: List[CommandParameter] = field(default_factory=list)


@dataclass
class DirectClientOptions:
    """Options for DirectClient."""
    
    source: Optional[str] = None
    debug: bool = False
    validate_inputs: bool = True


@dataclass
class DirectCallContext:
    """Context for individual command calls."""
    
    trace_id: Optional[str] = None
    timeout: Optional[float] = None
    extra: Dict[str, Any] = field(default_factory=dict)


# ═══════════════════════════════════════════════════════════════════════════════
# UNKNOWN TOOL ERROR
# ═══════════════════════════════════════════════════════════════════════════════


@dataclass
class UnknownToolError:
    """Structured error when an agent calls a non-existent tool."""
    
    error: str = "UNKNOWN_TOOL"
    message: str = ""
    requested_tool: str = ""
    available_tools: List[str] = field(default_factory=list)
    suggestions: List[str] = field(default_factory=list)
    hint: Optional[str] = None


def _levenshtein_similarity(a: str, b: str) -> float:
    """Calculate similarity between two strings (0 to 1)."""
    a_lower = a.lower()
    b_lower = b.lower()
    
    if a_lower == b_lower:
        return 1.0
    
    len_a, len_b = len(a_lower), len(b_lower)
    
    # Create distance matrix
    matrix = [[0] * (len_b + 1) for _ in range(len_a + 1)]
    
    for i in range(len_a + 1):
        matrix[i][0] = i
    for j in range(len_b + 1):
        matrix[0][j] = j
    
    for i in range(1, len_a + 1):
        for j in range(1, len_b + 1):
            cost = 0 if a_lower[i - 1] == b_lower[j - 1] else 1
            matrix[i][j] = min(
                matrix[i - 1][j] + 1,      # deletion
                matrix[i][j - 1] + 1,      # insertion
                matrix[i - 1][j - 1] + cost  # substitution
            )
    
    max_len = max(len_a, len_b)
    if max_len == 0:
        return 1.0
    return 1.0 - matrix[len_a][len_b] / max_len


def _find_similar_tools(
    requested: str,
    available: List[str],
    max_suggestions: int = 3
) -> List[str]:
    """Find similar tool names for suggestions."""
    scored = [
        (tool, _levenshtein_similarity(requested, tool))
        for tool in available
    ]
    filtered = [(t, s) for t, s in scored if s >= 0.4]
    filtered.sort(key=lambda x: x[1], reverse=True)
    return [t for t, _ in filtered[:max_suggestions]]


def _create_unknown_tool_error(
    requested: str,
    available: List[str]
) -> UnknownToolError:
    """Create a structured unknown tool error."""
    suggestions = _find_similar_tools(requested, available)
    hint = f"Did you mean '{suggestions[0]}'?" if suggestions else None
    
    return UnknownToolError(
        error="UNKNOWN_TOOL",
        message=f"Tool '{requested}' not found in registry",
        requested_tool=requested,
        available_tools=available,
        suggestions=suggestions,
        hint=hint,
    )


# ═══════════════════════════════════════════════════════════════════════════════
# DIRECT REGISTRY PROTOCOL
# ═══════════════════════════════════════════════════════════════════════════════


@runtime_checkable
class DirectRegistry(Protocol):
    """Protocol for registries that support direct execution.
    
    Implement this in your application to enable DirectClient.
    
    Example:
        >>> class MyRegistry:
        ...     def __init__(self):
        ...         self._commands = {}
        ...     
        ...     async def execute(self, name, args, context):
        ...         handler = self._commands.get(name)
        ...         return await handler(args)
        ...     
        ...     def has_command(self, name):
        ...         return name in self._commands
        ...     
        ...     def list_commands(self):
        ...         return [{'name': n, 'description': ''} for n in self._commands]
        ...     
        ...     def list_command_names(self):
        ...         return list(self._commands.keys())
    """
    
    async def execute(
        self,
        name: str,
        args: Optional[Dict[str, Any]] = None,
        context: Optional[CommandContext] = None,
    ) -> CommandResult:
        """Execute a command by name."""
        ...
    
    def has_command(self, name: str) -> bool:
        """Check if command exists."""
        ...
    
    def list_commands(self) -> List[CommandInfo]:
        """List all registered commands."""
        ...
    
    def list_command_names(self) -> List[str]:
        """List command names only."""
        ...
    
    # Optional method for validation support
    def get_command(self, name: str) -> Optional[CommandDefinition]:
        """Get command definition for validation (optional)."""
        ...


# ═══════════════════════════════════════════════════════════════════════════════
# SIMPLE REGISTRY IMPLEMENTATION
# ═══════════════════════════════════════════════════════════════════════════════


@dataclass
class RegisteredCommand:
    """A registered command with its handler and metadata."""
    
    name: str
    description: str
    handler: Callable
    parameters: List[CommandParameter] = field(default_factory=list)


class SimpleRegistry:
    """Simple in-memory command registry.
    
    Example:
        >>> registry = SimpleRegistry()
        >>> 
        >>> @registry.command(name='hello', description='Say hello')
        ... async def hello(name: str = 'World'):
        ...     return success({'message': f'Hello, {name}!'})
        >>> 
        >>> client = DirectClient(registry)
        >>> result = await client.call('hello', {'name': 'Agent'})
    """
    
    def __init__(self):
        self._commands: Dict[str, RegisteredCommand] = {}
    
    def command(
        self,
        name: str,
        description: str = "",
        parameters: Optional[List[CommandParameter]] = None,
    ) -> Callable:
        """Decorator to register a command.
        
        Args:
            name: Command name (use dot.notation for namespacing)
            description: What the command does
            parameters: Parameter definitions for validation
        """
        def decorator(func: Callable) -> Callable:
            self._commands[name] = RegisteredCommand(
                name=name,
                description=description or func.__doc__ or "",
                handler=func,
                parameters=parameters or [],
            )
            return func
        return decorator
    
    def register(
        self,
        name: str,
        handler: Callable,
        description: str = "",
        parameters: Optional[List[CommandParameter]] = None,
    ) -> None:
        """Register a command imperatively."""
        self._commands[name] = RegisteredCommand(
            name=name,
            description=description or handler.__doc__ or "",
            handler=handler,
            parameters=parameters or [],
        )
    
    async def execute(
        self,
        name: str,
        args: Optional[Dict[str, Any]] = None,
        context: Optional[CommandContext] = None,
    ) -> CommandResult:
        """Execute a command by name."""
        cmd = self._commands.get(name)
        if not cmd:
            return failure(not_found_error(f"Command '{name}' not found"))
        
        try:
            result = await cmd.handler(**(args or {}))
            return result
        except Exception as e:
            return failure(
                validation_error(f"Command execution failed: {e}")
            )
    
    def has_command(self, name: str) -> bool:
        """Check if command exists."""
        return name in self._commands
    
    def list_commands(self) -> List[CommandInfo]:
        """List all registered commands."""
        return [
            CommandInfo(name=cmd.name, description=cmd.description)
            for cmd in self._commands.values()
        ]
    
    def list_command_names(self) -> List[str]:
        """List command names only."""
        return list(self._commands.keys())
    
    def get_command(self, name: str) -> Optional[CommandDefinition]:
        """Get command definition for validation."""
        cmd = self._commands.get(name)
        if not cmd:
            return None
        return CommandDefinition(
            name=cmd.name,
            description=cmd.description,
            parameters=cmd.parameters,
        )


# ═══════════════════════════════════════════════════════════════════════════════
# DIRECT CLIENT
# ═══════════════════════════════════════════════════════════════════════════════


def _generate_trace_id() -> str:
    """Generate a unique trace ID."""
    return f"trace-{uuid.uuid4().hex[:12]}"


class DirectClient:
    """Zero-overhead command execution client.
    
    Calls commands directly via the registry without MCP transport.
    Provides ~0.1ms latency vs 10-100ms for MCP.
    
    Example:
        >>> client = DirectClient(registry)
        >>> result = await client.call('todo.create', {'title': 'Fast task'})
        >>> print(result.data)
    
    Pipeline Example:
        >>> result = await client.pipe([
        ...     {'command': 'user.get', 'input': {'id': 123}, 'as': 'user'},
        ...     {'command': 'order.list', 'input': {'user_id': '$user.id'}},
        ... ])
    """
    
    def __init__(
        self,
        registry: DirectRegistry,
        options: Optional[DirectClientOptions] = None,
    ):
        self.registry = registry
        self.options = options or DirectClientOptions()
    
    async def call(
        self,
        name: str,
        args: Optional[Dict[str, Any]] = None,
        context: Optional[DirectCallContext] = None,
    ) -> CommandResult:
        """Execute a command with zero transport overhead.
        
        Args:
            name: Command name
            args: Command arguments
            context: Optional call context for tracing
        
        Returns:
            CommandResult from the command handler
        """
        start_time = time.perf_counter() if self.options.debug else 0
        trace_id = (context.trace_id if context else None) or _generate_trace_id()
        
        self._debug(f"[{trace_id}] Calling {name}", args)
        
        # Check if command exists
        if not self.registry.has_command(name):
            available = self.registry.list_command_names()
            error_data = _create_unknown_tool_error(name, available)
            
            self._debug(f"[{trace_id}] Unknown command: {name}")
            
            return CommandResult(
                success=False,
                data=error_data,
                error={
                    "code": "UNKNOWN_TOOL",
                    "message": error_data.message,
                },
            )
        
        # Input validation (if registry supports getCommand)
        if self.options.validate_inputs and hasattr(self.registry, 'get_command'):
            cmd_def = self.registry.get_command(name)
            if cmd_def:
                issues = self._validate_input(args, cmd_def.parameters)
                if issues:
                    self._debug(f"[{trace_id}] Validation failed: {issues}")
                    return failure(
                        validation_error(
                            f"Invalid input for '{name}': {'; '.join(issues)}"
                        )
                    )
        
        # Build command context
        cmd_context = CommandContext(
            trace_id=trace_id,
            source=self.options.source,
            timeout=context.timeout if context else None,
            extra=context.extra if context else {},
        )
        
        # Execute the command
        result = await self.registry.execute(name, args, cmd_context)
        
        if self.options.debug:
            duration = (time.perf_counter() - start_time) * 1000
            self._debug(f"[{trace_id}] Completed in {duration:.3f}ms", {
                "success": result.success,
            })
        
        return result
    
    def list_commands(self) -> List[CommandInfo]:
        """List available commands."""
        return self.registry.list_commands()
    
    def list_command_names(self) -> List[str]:
        """List command names."""
        return self.registry.list_command_names()
    
    def has_command(self, name: str) -> bool:
        """Check if a command exists."""
        return self.registry.has_command(name)
    
    def get_source(self) -> Optional[str]:
        """Get the source identifier for this client."""
        return self.options.source
    
    def _validate_input(
        self,
        args: Optional[Dict[str, Any]],
        parameters: List[CommandParameter],
    ) -> List[str]:
        """Validate input against parameter definitions."""
        issues = []
        args = args or {}
        
        for param in parameters:
            if param.required and param.name not in args:
                issues.append(f"Missing required parameter: '{param.name}'")
                continue
            
            if param.name in args:
                value = args[param.name]
                # Type checking (basic)
                if param.type == "string" and not isinstance(value, str):
                    issues.append(f"'{param.name}' must be a string")
                elif param.type == "number" and not isinstance(value, (int, float)):
                    issues.append(f"'{param.name}' must be a number")
                elif param.type == "boolean" and not isinstance(value, bool):
                    issues.append(f"'{param.name}' must be a boolean")
                
                # Enum validation
                if param.enum and value not in param.enum:
                    issues.append(
                        f"'{param.name}' must be one of: {param.enum}"
                    )
        
        return issues
    
    def _debug(self, message: str, data: Any = None) -> None:
        """Debug logging helper."""
        if self.options.debug:
            print(f"[DirectClient] {message}", data if data else "")


# ═══════════════════════════════════════════════════════════════════════════════
# FACTORY FUNCTIONS
# ═══════════════════════════════════════════════════════════════════════════════


def create_registry() -> SimpleRegistry:
    """Create a new SimpleRegistry instance.
    
    Example:
        >>> registry = create_registry()
        >>> 
        >>> @registry.command(name='ping', description='Health check')
        ... async def ping():
        ...     return success({'status': 'ok'})
    """
    return SimpleRegistry()


def create_direct_client(
    registry: DirectRegistry,
    source: Optional[str] = None,
    debug: bool = False,
    validate_inputs: bool = True,
) -> DirectClient:
    """Create a DirectClient for zero-overhead command execution.
    
    Args:
        registry: The command registry to execute against
        source: Optional source identifier (e.g., 'my-agent')
        debug: Enable debug logging
        validate_inputs: Validate inputs against schemas
    
    Returns:
        Configured DirectClient instance
    
    Example:
        >>> client = create_direct_client(registry, source='api-server')
        >>> result = await client.call('user.get', {'id': 123})
    """
    return DirectClient(
        registry,
        DirectClientOptions(
            source=source,
            debug=debug,
            validate_inputs=validate_inputs,
        ),
    )
